#15 Create a Barcode

dmdwns2 opened the pull request:
ISSUE

https://github.com/a01575dc-b401-4506-8753-7d523b558dfa/membership-service/issues/15

To do

- [x] 회원 ID(9자리 숫자) 당 멤버십 ID(= 바코드 ID)가 10자리의 숫자 문자열로 발급됩니다.
- [x] 이미 발급된 아이디의 발급 요청 시 기존 멤버십 바코드를 반환합니다.
- [x] 다음에 발급되는 멤버십 바코드는 예측할 수 없어야 합니다.

Considerations

` N/A `

dmdwns2 committed:
findByuserId 메서드 추가

dmdwns2 committed:
- 유저ID로 바코드 ID가 있는지 확인 후 있으면 원래 있던 바코드 ID를 반환하는 코드 추가
- 코드 리팩토링

a01575dc-b401-4506-8753-7d523b558dfa commented on src/main/java/com/example/membership/service/infra/entity/MembershipRepository.java:
```diff
@@ -2,5 +2,8 @@
+import java.util.Optional;
+
 public interface MembershipRepository  extends JpaRepository<Membership, String> {
+    Optional<Membership> findByUserId(Long userId);
```
현재 인덱스 안 걸려있어서 full scan 을 돌 것 같음.
추후 schema 파일을 준비해서 이쪽에 정의하는 방향으로 정리하면 될듯.

a01575dc-b401-4506-8753-7d523b558dfa commented on src/main/java/com/example/membership/service/service/BarcodeCreateService.java:
```diff
@@ -0,0 +1,35 @@
+    private final UserRepository userRepository;
+    private final MembershipRepository membershipRepository;
+
+    public Membership createBarcode(Long userId) {
```
여기 user id 의 범위를 검증하면 좋을 것 같음.

a01575dc-b401-4506-8753-7d523b558dfa commented on src/main/java/com/example/membership/service/service/BarcodeCreateService.java:
```diff
@@ -0,0 +1,35 @@
+
+    public Membership createBarcode(Long userId) {
+        final User user = userRepository.findById(userId)
+                .orElseThrow(() -> new UserNotFoundException(userId));
```
`UserNotFoundException` 이 정의가 안 돼있는데, 아마 날아간 듯?

a01575dc-b401-4506-8753-7d523b558dfa commented on src/main/java/com/example/membership/service/service/BarcodeCreateService.java:
```diff
@@ -0,0 +1,35 @@
+        final User user = userRepository.findById(userId)
+                .orElseThrow(() -> new UserNotFoundException(userId));
+        return membershipRepository.findByUserId(user.getId())
+                .orElseGet(() -> saveAndGetBarcode(user));
```
동일한 user 에 대해서 생성 요청이 들어왔을 때,
둘 다 save 를 시도할 수 있을 것 같음.

나머지 1 개는 실패할텐데, 전체적으로 보면 성공한게 의미상 맞으므로
실패한 경우에 대해 예외를 잡아서 duplicate 인지 확인한다.
이 경우 다시 find 해서 반환해주면 좋을 것 같다.

dmdwns2 replied:
1. membership id 에 unique를 걸어서 해결하는 방식으로 했는데 이번에 릴레이션 다 삭제한다는거랑 관련이 있을지 모르겠음.
2. id 자체에 unique를 걸어서 중복이 나오면 jpa의 DataIntegrityViolationException 라는 데이터무결성에러를 catch로 잡는 식으로 해결했음.
3. db에서 무결성 에러가 나면 이미 있던 membership id를 반환하고 그래도 없으면 BarcodeSaveAndGetException 자체의 에러가 발생하게 처리해봄

dmdwns2 committed:
- membership id를 unique로 변경
- user id가 9자리인지 검증, 그에 대한 UserIdLengthException 추가
- 누락된 UserNotFoundException 추가
- 동일한 user가 동시에 saveAndGetBarcode 메서드를 실행한 경우 에러처리 추가
- BarcodeSaveAndGetException 추가

dmdwns2 committed:
- apache 추가

dmdwns2 commented:
피드백 반영 후 commit 완료
1. findByUserId 는 추후에 다시 정한다고해서 일단 냅뒀음
2. membership id에 unique 설정을 했는데 이것도 나중에 schema 설정할 때 체크해야될듯

a01575dc-b401-4506-8753-7d523b558dfa commented on src/main/java/com/example/membership/service/infra/entity/Membership.java:
```diff
@@ -2,13 +2,16 @@
 @Entity
 public class Membership {
     @Id
+    @Column(unique = true)
```
이건 제거해도 될 것 같다.

a01575dc-b401-4506-8753-7d523b558dfa commented on src/main/java/com/example/membership/service/service/BarcodeCreateService.java:
```diff
@@ -0,0 +1,42 @@
+    public Membership createBarcode(Long userId) {
+        final User user = userRepository.findById(userId)
+                .orElseThrow(() -> new UserNotFoundException(userId));
+        if(!user.getId().toString().matches("\\d{9}")){
+            throw new UserIdLengthException(userId);
+        }
```
여기는 검증을 실제 조회전에 하는 것으로 변경하는게 좋을 것 같다.
이미 존재하는 entity 는 따로 검증하지 않아도 좋을듯.

dmdwns2 committed:
- @Column의 unique 를 명시할 필요가 없어 삭제
- input을 가장먼저 검증하도록 service 변경
- 테스트코드 오타 수정

a01575dc-b401-4506-8753-7d523b558dfa commented on src/main/java/com/example/membership/service/service/BarcodeCreateService.java:
```diff
@@ -0,0 +1,42 @@
+        final User user = userRepository.findById(userId)
+                .orElseThrow(() -> new UserNotFoundException(userId));
+        try {
+            return membershipRepository.findByUserId(user.getId())
```
현재는 full table scan 을 타고 있는데,
https://github.com/a01575dc-b401-4506-8753-7d523b558dfa/membership-service/issues/17 진행하면서 인덱스를 걸자.

a01575dc-b401-4506-8753-7d523b558dfa commented on src/main/java/com/example/membership/service/service/BarcodeCreateService.java:
```diff
@@ -0,0 +1,42 @@
+    private final MembershipRepository membershipRepository;
+
+    public Membership createBarcode(Long userId) {
+        if(!userId.toString().matches("\\d{9}")){
```
지금은 length 비교만 하면 돼서, regex 를 굳이 안 써도 될 것 같다.
regex 는 호불호가 있기도 하고, 상대적으로 수행시간이 길기 때문에 변경하면 좋을듯

dmdwns2 committed:
- userID의 값이 null, 음수인 경우 검증, 가독성 향상

a01575dc-b401-4506-8753-7d523b558dfa commented on src/main/java/com/example/membership/service/service/BarcodeCreateService.java:
```diff
@@ -0,0 +1,52 @@
+
+@Service
+@RequiredArgsConstructor
+@Transactional
```
여기 transaction 이 필요 없어보인당.

dmdwns2 committed:
- service에 트랜잭션 삭제
- BarcodeSaveAndGetException 에러 설명을 명확하게 변경

a01575dc-b401-4506-8753-7d523b558dfa reviewed (approved):
고생하셨습니다. 🙇
